File: 0169-majority-element/0169-majority-element.py
from typing import List


class Solution:

    def count(self, a: List[int], x: int) -> int:
        ans = 0
        for i in a:
            if i == x:
                ans += 1
        return ans

    def majorityElement(self, nums: List[int]) -> int:
        n = len(nums)

        if n == 1:
            return nums[0]

        x = nums[: n // 2]
        y = nums[n // 2 :]

        l = self.majorityElement(x)
        r = self.majorityElement(y)

        cl = self.count(nums, l)
        cr = self.count(nums, r)

        if cl > n / 2:
            return l
        if cr > n / 2:
            return r

        return -1

File: 0169-majority-element/0169-majority-element.cpp
class Solution {
public:
    int majorityElement(vector<int>& nums) {
        int cnt = 0;
        int n = nums.size();
        int ans = 0;
        for (int i = 0; i < n; ++i) {
            if (cnt == 0) {
                ans = nums[i];
            }

            if (ans != nums[i]) {
                cnt--;
            }else {
                cnt++;
            }
        }
        return ans;
    }
};